TransformationPlayground/playground/transform_expert/expert.py

```diff
@@ -0,0 +1,139 @@
+        max_tokens=4096,
+        region_name="us-west-2", # Somewhat necessary to hardcode, as models are only available in limited regions
+        config=config
+    )
+    llm_w_tools = llm.bind_tools(tool_bundle.to_list())
+
+    return Expert(
+        llm=llm_w_tools,
+        system_prompt_factory=get_system_prompt_factory(
+            source_version=source_version,
+            target_version=target_version,
+            input_shape_type=transform_type,
+            transform_language=transform_language
+        ),
+        tools=tool_bundle
+    )
+
+def invoke_expert(expert: Expert, task: TransformTask) -> TransformTask:
+    logger.info(f"Invoking the Transform Expert for transform_id: {task.transform_id}")
+    logger.debug(f"Initial Transform Task: {str(task.to_json())}")
+
+    # Invoke the LLM.  This should result in the LLM making a tool call, forcing it to create the transform details by
+    # conforming to the tool's schema.
+    inference_task = InferenceTask.from_transform_task(task)
+    inference_result = perform_inference(expert.llm, [inference_task])[0]
+
+    logger.debug(f"Inference Result: {str(inference_result.to_json())}")
+
+    # Append the LLM response to the context
+    task.context.append(inference_result.response)
+
+    # Perform the tool call using the LLM's response to create our Transform object
+    tool_call = inference_result.response.tool_calls[-1]
+    transform = expert.tools.make_transform_tool(tool_call["args"])
+    task.transform = transform
+
+    # Append the tool call to the context
+    task.context.append(
+        ToolMessage(
+                name="MakePythonTransform",
+                content="Created the transform",
+                tool_call_id=tool_call["id"]
+        )
+    )
+
+    logger.info(f"Transform completed for transform_id: {task.transform_id}")
+    logger.debug(f"Updated Transform Task: {str(task.to_json())}")
+
+    return task
+
+
+@dataclass
+class InferenceTask:
+    transform_id: str
+    context: List[BaseMessage]
+
+    @staticmethod
+    def from_transform_task(task: TransformTask) -> 'InferenceTask':
+        return InferenceTask(
+            transform_id=task.transform_id,
+            context=task.context
+        )
+
+    def to_json(self) -> dict:
+        return {
+            "transform_id": self.transform_id,
+            "context": [turn.to_json() for turn in self.context]
+        }
+
+@dataclass
+class InferenceResult:
+    transform_id: str
+    response: BaseMessage
+
+    def to_json(self) -> dict:
+        return {
+            "transform_id": self.transform_id,
+            "response": self.response.to_json()
+        }
+
+
+def perform_inference(llm: Runnable[LanguageModelInput, BaseMessage], batched_tasks: List[InferenceTask]) -> List[InferenceResult]:
+    return asyncio.run(_perform_async_inference(llm, batched_tasks))
+
+# Inference APIs can be throttled pretty aggressively.  Performing them as a batch operation can help with increasing
```

Comment:
That's a good question.  Basically, it depends on how your split up the overall work of inference.  In our scenario, I expect this batching to be used when the user wants to generate many different transforms simultaneously; think about them loading in a bunch of indices and hitting a general "gimme recommendations" button. Each index would have its own LLM conversation, etc.  To support this, the Django API would need a revamp to handle these sorts of batch operations, among other things.

I have this async code in here currently because it was a life-saver on some of my prototyping efforts (remember the document summarizer?) that required splitting up the overall workload into many chunks due to limitations around the number of output tokens I could receive back for each chunk.